<launch>

    <arg name="result_file_name" default="$(find hiltislamchallenge)/output/result"/>
    <arg name="bagfilename" default="/home/mbo/norlab_data/hilti/long_corridor/bagfiles/exp07_long_corridor.bag">

    <!-- either "complementary", or "madgwick"    -->
    <arg name="imu_filter" default="complementary"/>

    <node pkg="norlab_icp_mapper_ros" type="mapper_node" name="mapper_node" output="screen" respawn="false">
        <remap from="points_in" to="/hesai/pandar_deskewed"/>
        <param name="robot_frame" value="base_link"/>
        <param name="odom_frame" value="odom"/>

        <param name="icp_config" value="$(find hiltislamchallenge)/params/realtime_icp_config.yaml"/>
        <param name="input_filters_config" value="$(find hiltislamchallenge)/params/realtime_input_filters.yaml"/>
        <param name="map_post_filters_config" value="$(find hiltislamchallenge)/params/realtime_post_filters.yaml"/>

        <param name="map_update_condition" value="delay"/>
        <param name="map_update_delay" value="0"/>

        <param name="map_publish_rate" value="10" />
        <param name="map_tf_publish_rate" value="100" />
        <param name="max_idle_time" value="10" />
        <param name="min_dist_new_point" value="0.03" />
        <param name="sensor_max_range" value="100" />
        <param name="prior_dynamic" value="0.5" />
        <param name="threshold_dynamic" value="0.75" />
        <param name="beam_half_angle" value="0.003054326" />
        <param name="epsilon_a" value="0.01" />
        <param name="epsilon_d" value="0.01" />
        <param name="alpha" value="0.8" />
        <param name="beta" value="0.99" />
        <param name="is_3D" value="true" />
        <param name="is_online" value="true" />
        <param name="compute_prob_dynamic" value="True" />
        <param name="is_mapping" value="true" />

        <param name="final_map_file_name" value="$(arg result_file_name)_final_map.vtk"/>
        <param name="final_trajectory_file_name" value="$(arg result_file_name)_final_traj.vtk"/>
    </node>

    <param name="robot_description" command="$(find xacro)/xacro '$(find hiltislamchallenge)/urdf/hilti.urdf.xacro'"/>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>

    <node pkg="hiltislamchallenge" type="imu_bias_compensation" name="imu_bias_compensation" output="screen">
        <remap from="/imu/data_in" to="/alphasense/imu"/>
        <remap from="/imu/data_out" to="/alphasense/imu_compensated"/>

        <rosparam param="linear_velocity_bias">[0.0010687901823426531, -0.0005818020879880394, -0.00008079731341682668]
        </rosparam>
    </node>

    <group if="$(eval arg('imu_filter')=='complementary')">
        <node pkg="imu_complementary_filter" type="complementary_filter_node" name="complementary_filter_gain_node" output="screen">
            <param name="do_bias_estimation" value="false"/>
            <param name="do_adaptive_gain" value="true"/>
            <param name="use_mag" value="false"/>
            <param name="gain_acc" value="0.01"/>
            <param name="gain_mag" value="0.01"/>
            <param name="publish_debug_topics" value="false"/>
            <param name="publish_tf" value="false"/>
            <param name="odom_frame" value="odom" />
            <param name="base_frame" value="base_link" />

            <remap from="/imu/data_raw" to="/alphasense/imu_compensated"/>
            <remap from="/imu/data" to="/alphasense/imu_filtered"/>

        </node>
    </group>

    <group if="$(eval arg('imu_filter')=='madgwick')">
        <node pkg="imu_filter_madgwick" type="imu_filter_node" name="madgwick_filter_node" output="screen">
            <param name="gain" value="0.1"/>
            <param name="zeta" value="0.0"/>
            <param name="world_frame" value="enu"/>
            <param name="use_mag" value="false"/>
            <param name="use_magnetic_field_msg" value="false"/>
            <param name="fixed_frame" value="odom"/>
            <param name="publish_tf" value="false"/>
            <param name="publish_debug_topics" value="false"/>

            <remap from="/imu/data_raw" to="/alphasense/imu_compensated"/>
            <remap from="/imu/data" to="/alphasense/imu_filtered"/>

        </node>
    </group>

    <node pkg="imu_odom" type="imu_odom_node" name="imu_odom_node" output="screen">
        <remap from="imu_topic" to="/alphasense/imu_filtered"/>
        <param name="odom_frame" value="odom"/>
        <param name="robot_frame" value="base_link"/>
        <param name="imu_frame" value="imu_sensor_frame"/>
        <param name="real_time" value="true"/>
    </node>

    <node name="pcl_deskew_node" pkg="pointcloud_motion_deskew" type="pointcloud2_deskew_node" output="log">
        <remap from="input_point_cloud" to="/hesai/pandar" />
        <remap from="output_point_cloud" to="/hesai/pandar_deskewed" />
    </node>

    <node pkg="hiltislamchallenge" type="mapping_monitor.py" name="mapping_monitor" output="screen">
        <param name="bagfile" value="$(arg bagfilename)"/>
    </node>

</launch>
